backend_new/src/services/user.service.ts

```diff
@@ -0,0 +1,234 @@
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { PrismaService } from './prisma.service';
+import { User } from '../dtos/users/user.dto';
+import { mapTo } from '../utilities/mapTo';
+import {
+  buildPaginationMetaInfo,
+  calculateSkip,
+  calculateTake,
+} from '../utilities/pagination-helpers';
+import { buildOrderBy } from '../utilities/build-order-by';
+import { Prisma } from '@prisma/client';
+import { UserQueryParams } from '../dtos/users/user-query-param.dto';
+import { PaginatedUserDto } from '../dtos/users/paginated-user.dto';
+import { OrderByEnum } from '../enums/shared/order-by-enum';
+import { UserFilterKeys } from '../enums/user_accounts/filter-key-enum';
+
+/*
+  this is the service for users
+  it handles all the backend's business logic for reading/writing/deleting user data
+*/
+
+const view: Prisma.UserAccountsInclude = {
+  listings: true,
+  jurisdictions: true,
+  userRoles: true,
+};
+
+@Injectable()
+export class UserService {
+  constructor(private prisma: PrismaService) {}
+
+  /*
+    this will get a set of users given the params passed in
+  */
+  async list(params: UserQueryParams, user: User): Promise<PaginatedUserDto> {
+    const whereClause = this.buildWhereClause(params, user);
+    const count = await this.prisma.userAccounts.count({
+      where: whereClause,
+    });
+
+    const rawUsers = await this.prisma.userAccounts.findMany({
+      skip: calculateSkip(params.limit, params.page),
+      take: calculateTake(params.limit),
+      orderBy: buildOrderBy(
+        ['firstName', 'lastName'],
+        [OrderByEnum.ASC, OrderByEnum.ASC],
+      ),
+      include: view,
+      where: whereClause,
+    });
+
+    const users = mapTo(User, rawUsers);
+
+    const paginationInfo = buildPaginationMetaInfo(params, count, users.length);
+
+    return {
+      items: users,
+      meta: paginationInfo,
+    };
+  }
+
+  /*
+    this helps build the where clause for the list()
+  */
+  buildWhereClause(
+    params: UserQueryParams,
+    user: User,
+  ): Prisma.UserAccountsWhereInput {
+    const filters: Prisma.UserAccountsWhereInput[] = [];
+
+    if (params.search) {
```

Comment:
Same suggestion as past PRs with updating to return if no params, instead of wrapping this in a conditional.

Reply:
Hey Sean I do that below this line with `if (!params.filter?.length)` we just return otherwise we go through the param filter's array and build each filter

for this case params has an optional search and an optional filter so I think we need this if statement to capture if params.search exists then we build the filter for it

even if we did that opposite I think we would basically have the same thing with 

if (!params.search) { // do nothing} < insert the building of the search filter here> 

hopefully that makes some sense?